### ISSUE TITLE
Transfer learning

### ISSUE THREAD

**Issue by NVukobrat** [ZHKKKe/MODNet#26]
Hi,

Do you plan to release transfer learning options to specify which layers should be trained? 

Also, have you tried this network on different objects besides portraits? If yes, did you managed to get comparable results?

**Comment by ZHKKKe** [comment]
Hi, thanks for your attention.

For your questions:

**Q1: Do you plan to release transfer learning options to specify which layers should be trained?**
Does "transfer learning" mean the SOC strategy in our paper? If so, we will provide training code for the SOC stage.

**Q2: Also, have you tried this network on different objects besides portraits? If yes, did you managed to get comparable results?**
We didn't try it. As far as I know, the only open-source matting dataset that consists of different objects is Adobe Matting Dataset. However, this dataset is too small to train a trimap-free matting model. 
If you focus on a specific object, e.g., cars, and you have a labeled dataset, I believe you can get results comparable to our portrait matting scheme.

**Comment by 99991** [comment]
I've tried the network on random images and it seems to work quite well. Best results can be achieved if the subject has a certain distance to the background. My hypothesis is that the network might be conditioned to cut out objects which are not in focus. For example, in many images, the background is out of focus.

![4x4 grid of various images](https://user-images.githubusercontent.com/18725165/102453605-d757c100-403c-11eb-9f63-a0c2c15f997e.jpeg)

**Comment by ZHKKKe** [comment]
@99991 

Wowoo... This is a very interesting finding. It means that the model has learned the pattern of bokeh.
I agree with you. If the background is not in focus, i.e., bokeh, the network will find the objects that are in focus, i.e., the saliency objects. If the background is also in focus, the network will try to find the portraits.

**Comment by NVukobrat** [comment]
Thanks for the answers! @ZHKKKe @99991 

**Q1: Does "transfer learning" mean the SOC strategy in our paper?**
I wasn't thinking about your SOC approach here. Rather about will your training code provide an option to load a pre-trained model and then fine-tune specific layers of the network while freezing others? 

**Q3: Do you know what is training time on certain hardware for your model? Eather train from scratch or by doing transfer learning if applicable.** 

Btw, really great results guys on this one! I can't wait to see how this network will perform on different types of objects and backgrounds 😄 

Also, great results @99991. Before asking this question, I also tested the model on different objects without a "bokeh" background but rather full image focus. As @ZHKKKe said, the results weren't pretty good as the network cut out the whole background certain parts of the desired object too.

**Comment by ZHKKKe** [comment]
@NVukobrat 

**Q1: Does "transfer learning" mean the SOC strategy in our paper?**
We will provide the training process, including the loading function of the pre-trained model. However, since there are too many layers in the model, we will not add options to freeze certain layer parameters. Therefore, it is necessary for code users to add extra functions to achieve this (I believe this is not difficult).

**Q3: Do you know what is training time on certain hardware for your model? Eather train from scratch or by doing transfer learning if applicable.**
Our supervised training set contains about 100k samples (3k foregrounds + background replacement). It will take 2-3 days in two GTX 1080Ti GPUs. Our SOC strategy uses about 50k unlabeled images. It will take 1-2 days in a single GTX 1080Ti GPU.

**Comment by newjavaer** [comment]
@ZHKKKe Does the dataset will be public?

**Comment by ZHKKKe** [comment]
@newjavaer 
Nope. I am sorry that we cannot release the training set due to license issue.

**Comment by NVukobrat** [comment]
Thanks again @ZHKKKe for your answers! 

**Q4: Do you have any timeline for when can we expect the training script?** 

**Q5: Also, can you give us insights into which dataset structure training script will support? For example, should data be split into different directories in some way, should labels be in COCO or some other format... Do you have more info on this one?**

**Comment by ZHKKKe** [comment]
@NVukobrat 

**Q4: Do you have any timeline for when can we expect the training script?**
We will release the training code next month.

**Q5: Also, can you give us insights into which dataset structure training script will support? For example, should data be split into different directories in some way, should labels be in COCO or some other format... Do you have more info on this one?**
We will provide a superclass (template) for the code of dataset. You can inherit from it and implement your own dataset. 
We may provide a default dataset class for Adobe Image Matting Dataset.

**Comment by NVukobrat** [comment]
Great! Thanks for the help @ZHKKKe!
